```{r}
library(Seurat)
library(Signac)
library(ggplot2)

library(rtracklayer)
library(GenomicRanges)
library(dplyr)

base <- "/media/chang/HDD-8/joseph/bmi206"
```

# exons and cis non-coding

```{r}
# extract from annotation GTF
# gtf <- rtracklayer::import(file.path(base, "gencode.v47.annotation.gtf.gz"))

# # all genes
# genes <- gtf[gtf$type == "gene"]
# saveRDS(genes, file = file.path(base, "genes.rds"))

# exons <- gtf[gtf$type == "exon"]

# # all exons genome-wide
# all_exons_union <- reduce(exons)
# saveRDS(all_exons_union, file = file.path(base, "all_exons_union.rds"))

# # exons by gene
# exons_by_gene <- split(exons, exons$gene_id)
# exonic_regions_by_gene <- reduce(exons_by_gene)
# saveRDS(exonic_regions_by_gene, file = file.path(base, "exonic_regions_by_gene.rds"))

genes <- readRDS(file.path(base, "genes.rds"))
# a single gene_name can map to multiple gene_id (different isoforms), so use gene_id when known

exonic_regions_by_gene <- readRDS(file.path(base, "exonic_regions_by_gene.rds"))
all_exons_union <- readRDS(file.path(base, "all_exons_union.rds"))

pos <- ifelse(strand(genes) == "+", start(genes), end(genes))
tss <- GRanges(seqnames(genes), IRanges(pos, pos), strand = strand(genes), gene_id = genes$gene_id, gene_name = genes$gene_name)

phast_bw <- file.path(base, "hg38.phastCons100way.bw")
cis_window <- 500000L # 500 kb
```

```{r}
get_gene_phastcons_exon <- function(gene_id) {
  print(paste(which(gene_id == gene_ids), "/", length(gene_ids)))

  exons_g <- exonic_regions_by_gene[[gene_id]]
  if (is.null(exons_g) || length(exons_g) == 0L) {return(NULL)}
  
  pc_exon <- import(phast_bw, which = exons_g)
  if (length(pc_exon) == 0L) {return(NULL)}
  
  mean_exon <- mean(pc_exon$score, na.rm = TRUE)
  data.frame(gene_id = gene_id, mean_exon = mean_exon)
}

get_gene_phastcons_cis <- function(gene_id) {
  print(paste(which(gene_id == gene_ids), "/", length(gene_ids)))
  
  # gene TSS
  tss_g <- tss[tss$gene_id == gene_id]
  if (length(tss_g) == 0L) {return(NULL)}
  tss_g <- tss_g[1] # when there are multiple isoforms, just use the first one (if using gene_names)
  
  # gene cis window: [TSS - 500kb, TSS + 500kb]
  chr <- as.character(seqnames(tss_g))
  pos <- start(tss_g)
  
  # get noncoding (non-exonic) regions only
  cis_range <- GRanges(seqnames = chr, IRanges(start = max(1L, pos - cis_window), end = pos + cis_window))
  cis_noncoding <-  GenomicRanges::setdiff(cis_range, all_exons_union)
  if (length(cis_noncoding) == 0L) {return(NULL)}

  pc_cis  <- import(phast_bw, which = cis_noncoding)
  if (length(pc_cis) == 0L) {return(NULL)}
  
  mean_cis  <- mean(pc_cis$score,  na.rm = TRUE)
  data.frame(gene_id = gene_id, mean_cis = mean_cis)
}

gene_ids <- unique(genes$gene_ids)

gene_results_exon_list <- lapply(gene_ids, get_gene_phastcons_exon)
gene_results_exon <- do.call(rbind, gene_results_exon_list)
gene_results_exon <- gene_results_exon %>% mutate(gene_name = genes$gene_name[match(gene_id, genes$gene_id)])

gene_results_cis_list <- lapply(gene_ids, get_gene_phastcons_cis)
gene_results_cis   <- do.call(rbind, gene_results_cis_list)
gene_results_cis <- gene_results_cis %>% mutate(gene_name = genes$gene_name[match(gene_id, genes$gene_id)])

saveRDS(gene_results_exon, file = file.path(base, "gene_phastcons_exon.rds"))
saveRDS(gene_results_cis, file = file.path(base, "gene_phastcons_cis.rds"))

gene_results_exon <- readRDS(file.path(base, "gene_phastcons_exon.rds"))
gene_results_cis  <- readRDS(file.path(base, "gene_phastcons_cis.rds"))
```

# SCENT and ATAC peaks
```{r}
make_scent_peaks_by_gene <- function(scent_peaks) {
  # parse "peak" column from scent peak table
  scent_peaks <- scent_peaks %>% mutate(chrom = sub("-.*", "", peak), start = as.integer(sub(".*-(\\d+)-.*", "\\1", peak)), 
                                                                      end = as.integer(sub(".*-(\\d+)$", "\\1", peak)))
  # create GRanges object for each gene from scent_peaks
  scent_peaks_by_gene <- split(GRanges(seqnames = scent_peaks$chrom, 
                                      ranges = IRanges(start = scent_peaks$start, end = scent_peaks$end), 
                                      strand = "*"), 
                              scent_peaks$gene)
  return(scent_peaks_by_gene)
}

make_atac_peaks_by_gene <- function(atac_peaks, gene_names) { # every atac peak is ~200bp
  atac_peaks_by_gene <- list()
  for (gene_name in gene_names) {
    print(paste(which(gene_name == gene_names), "/", length(gene_names)))
    tss_g <- tss[tss$gene_name == gene_name]
    if (length(tss_g) == 0L) {next} # gene name not in gene reference
    tss_g <- tss_g[1] # when there are multiple isoforms, just use the first one

    # gene cis window: [TSS - 500kb, TSS + 500kb]
    cis_window <- 500000L
    chr <- as.character(seqnames(tss_g))
    pos <- start(tss_g)
    atac_peaks_by_gene[[gene_name]] <- atac_peaks[seqnames(atac_peaks) == chr &
                                                  start(atac_peaks) >= (pos - cis_window) &
                                                  end(atac_peaks)   <= (pos + cis_window)]
  }
  return(atac_peaks_by_gene)
}

get_gene_phastcons_scent <- function(gene_name, scent_peaks_by_gene) {
  print(paste(which(gene_name == scent_gene_names), "/", length(scent_gene_names)))

  peaks_g <- scent_peaks_by_gene[[gene_name]]

  pc_peaks <- import(phast_bw, which = peaks_g)
  if (length(pc_peaks) == 0L) {return(NULL)}
  
  mean_peaks <- mean(pc_peaks$score, na.rm = TRUE)

  data.frame(gene_name = gene_name, mean_scent = mean_peaks)
}

get_gene_phastcons_atac <- function(gene_name, atac_peaks_by_gene) {
  print(paste(which(gene_name == scent_gene_names), "/", length(scent_gene_names)))

  peaks_g <- atac_peaks_by_gene[[gene_name]]
  if (is.null(peaks_g) || length(peaks_g) == 0L) {return(NULL)}

  pc_peaks <- import(phast_bw, which = peaks_g)
  if (length(pc_peaks) == 0L) {return(NULL)}
  
  mean_peaks <- mean(pc_peaks$score, na.rm = TRUE)

  data.frame(gene_name = gene_name, mean_scent = mean_peaks)
}
```

```{r}
# dataset <- "Arthritis" # GSE243917
# scent_peaks_filepath <- file.path(base, dataset, "Arthritis-tissue_allqced_bootpkg_nopc_allCT.FDR0.10.txt") 
# seurat_obj <- readRDS(file.path(base, dataset, "seurat_arthritis.rds"))

# dataset <- "PBMC" # 10x Genomics
# scent_peaks_filepath <- file.path(base, dataset, "Pubic-PBMC_allqced_bootpkg_nopc_allCT.FDR0.10.txt")
# seurat_obj <- readRDS(file.path(base, dataset, "seurat_pbmc.rds"))

dataset <- "NeurIPS"
scent_peaks_filepath <- file.path(base, dataset, "NeurIPS_allqced_bootpkg_nopc_allCT.FDR0.10.txt")
seurat_obj <- readRDS(file.path(base, dataset, "seurat_neurips.rds"))

scent_peaks <- read.table(scent_peaks_filepath, header = TRUE, stringsAsFactors = FALSE)
scent_gene_names <- unique(scent_peaks$gene)

# gene_ids <- genes$gene_id[match(scent_gene_names, genes$gene_name)]
# sum(!(gene_ids %in% genes$gene_id))
# sum(!(scent_gene_names %in% genes$gene_name)) # number of genes from scent peaks not in gene_name of genes


## phastcons for SCENT peaks

scent_peaks_by_gene <- make_scent_peaks_by_gene(scent_peaks)

# compute conservation scores for each gene
gene_results_scent_list <- list()
for (i in seq_along(scent_gene_names)) {gene_results_scent_list[[i]] <- get_gene_phastcons_scent(scent_gene_names[i], scent_peaks_by_gene)}
gene_results_scent <- do.call(rbind, gene_results_scent_list)
saveRDS(gene_results_scent, file = file.path(base, dataset, "gene_phastcons_scent.rds"))


## phastcons for ATAC peaks

# atac_peaks <- StringToGRanges(Features(seurat_obj[["ATAC"]]), sep = c(":", "-")) # every range has at least one ATAC peak in seurat object
# atac_peaks_by_gene <- make_atac_peaks_by_gene(atac_peaks, scent_gene_names)
# saveRDS(atac_peaks_by_gene, file = file.path(base, dataset, "atac_peaks_by_gene.rds"))

atac_peaks_by_gene <- readRDS(file.path(base, dataset, "atac_peaks_by_gene.rds"))

# compute conservation scores for each gene
gene_results_atac_list <- list()
for (i in seq_along(scent_gene_names)) {gene_results_atac_list[[i]] <- get_gene_phastcons_atac(scent_gene_names[i], atac_peaks_by_gene)}
gene_results_atac <- do.call(rbind, gene_results_atac_list)
saveRDS(gene_results_atac, file = file.path(base, dataset, "gene_phastcons_atac.rds"))
```

# bootstrapping
```{r}
bootstrap_score <- function(delta_pc, n_boot = 100) {
  set.seed(1)
  boot_means <- replicate(n_boot, {mean(sample(delta_pc[!is.na(delta_pc)], size = sum(!is.na(delta_pc)), replace = TRUE))})
  ci <- quantile(boot_means, probs = c(0.025, 0.975), na.rm = TRUE) # 95% CI
  list(mean = mean(boot_means), ci_lower = ci[1], ci_upper = ci[2])
}

bootstraps_scent <- list()
bootstraps_atac <- list()

gene_pc_noncoding  <- readRDS(file.path(base, "gene_phastcons_cis.rds"))

datasets <- c("Arthritis", "PBMC", "NeurIPS")

all_scent_genes <- c()
for (dataset in datasets) {
  gene_pc_scent <- readRDS(file = file.path(base, dataset, "gene_phastcons_scent.rds"))
  gene_pc_atac  <- readRDS(file = file.path(base, dataset, "gene_phastcons_atac.rds"))

  delta_pc_scent <- gene_pc_scent$mean_scent - gene_pc_noncoding$mean_cis[match(gene_pc_scent$gene_name, gene_pc_noncoding$gene_name)]
  bootstraps_scent[[dataset]] <- bootstrap_score(delta_pc_scent)
  delta_pc_atac <- gene_pc_atac$mean_scent - gene_pc_noncoding$mean_cis[match(gene_pc_atac$gene_name, gene_pc_noncoding$gene_name)]
  bootstraps_atac[[dataset]] <- bootstrap_score(delta_pc_atac)

  all_scent_genes <- union(all_scent_genes, gene_pc_scent$gene_name)
}

gene_pc_exon <- readRDS(file.path(base, "gene_phastcons_exon.rds"))
gene_pc_exon <- gene_pc_exon %>% filter(gene_name %in% all_scent_genes)
delta_pc_exon <- gene_pc_exon$mean_exon - gene_pc_noncoding$mean_cis[match(gene_pc_exon$gene_id, gene_pc_noncoding$gene_id)]
bootstrap_exon <- bootstrap_score(delta_pc_exon)
```

# plotting
```{r}
bootstrap_results <- data.frame(
  Dataset = factor(c("Exons", "Arthritis tissue", "Arthritis tissue", "Public PBMC", "Public PBMC", "NeurIPS", "NeurIPS"), levels = c("Exons", "Arthritis tissue", "Public PBMC", "NeurIPS")),
  Category = factor(c("Exonic regions", "SCENT", "All cis-ATAC peaks", "SCENT", "All cis-ATAC peaks", "SCENT", "All cis-ATAC peaks"), levels = c("Exonic regions", "SCENT", "All cis-ATAC peaks")),
  Mean = c(bootstrap_exon$mean, bootstraps_scent[["Arthritis"]]$mean, bootstraps_atac[["Arthritis"]]$mean, bootstraps_scent[["PBMC"]]$mean, bootstraps_atac[["PBMC"]]$mean, bootstraps_scent[["NeurIPS"]]$mean, bootstraps_atac[["NeurIPS"]]$mean),
  CI_Lower = c(bootstrap_exon$ci_lower, bootstraps_scent[["Arthritis"]]$ci_lower, bootstraps_atac[["Arthritis"]]$ci_lower, bootstraps_scent[["PBMC"]]$ci_lower, bootstraps_atac[["PBMC"]]$ci_lower, bootstraps_scent[["NeurIPS"]]$ci_lower, bootstraps_atac[["NeurIPS"]]$ci_lower),
  CI_Upper = c(bootstrap_exon$ci_upper, bootstraps_scent[["Arthritis"]]$ci_upper, bootstraps_atac[["Arthritis"]]$ci_upper, bootstraps_scent[["PBMC"]]$ci_upper, bootstraps_atac[["PBMC"]]$ci_upper, bootstraps_scent[["NeurIPS"]]$ci_upper, bootstraps_atac[["NeurIPS"]]$ci_upper)
)

# Create the dot plot
p <- ggplot(bootstrap_results, aes(x = Dataset, y = Mean, color = Category)) +
  geom_point(size = 2, position = position_dodge(width = 0.25)) +
  geom_errorbar(aes(ymin = CI_Lower, ymax = CI_Upper), width = 0, position = position_dodge(width = 0.25)) +
  labs(y = "âˆ†phastCons score from all cis-regions") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, color = "black"), axis.text.y = element_text(color = "black"), axis.title.x = element_blank(), panel.grid = element_blank(), panel.border = element_blank(), axis.line = element_line(linewidth=0.2), axis.ticks = element_line(linewidth=0.2)) +
  scale_color_manual(values = c("Exonic regions" = "#6c488c", "SCENT" = "#00b1a4", "All cis-ATAC peaks" = "#e6be6f")) +
  ylim(-0.2, 0.4) +
  theme(legend.title = element_blank(), legend.position = "top") +
  geom_hline(yintercept = 0, color = "black", linewidth=0.2)
ggsave(p, filename = file.path("bmi206/joseph/result.png"), width = 5, height = 4)
```

# old code below

```{r}
dataset <- "Arthritis" # GSE243917
filepath <- file.path(base, dataset, "Arthritis-tissue_allqced_bootpkg_nopc_allCT.FDR0.10.txt")
scent_peaks <- read.table(filepath, header = TRUE, stringsAsFactors = FALSE)
scent_gene_names <- unique(scent_peaks$gene)

## SCENT peaks

# 
# gene_ids <- genes$gene_id[match(scent_gene_names, genes$gene_name)]
# sum(!(gene_ids %in% genes$gene_id))
# sum(!(scent_gene_names %in% genes$gene_name))

scent_peaks_by_gene <- make_scent_peaks_by_gene(scent_peaks)

# compute conservation scores for each gene
gene_results_scent_list <- list()
for (i in seq_along(scent_gene_names)) {gene_results_scent_list[[i]] <- get_gene_phastcons_scent(scent_gene_names[i], scent_peaks_by_gene)}
gene_results_scent <- do.call(rbind, gene_results_scent_list)
saveRDS(gene_results_scent, file = file.path(base, dataset, "gene_phastcons_scent.rds"))

## ATAC peaks

seurat_arthritis <- readRDS(file.path(base, dataset, "seurat_arthritis.rds"))
atac_peaks <- StringToGRanges(Features(seurat_arthritis), sep = c(":", "-")) # every range has at least one ATAC peak in seurat object
atac_peaks_by_gene <- make_atac_peaks_by_gene(atac_peaks, scent_gene_names)
saveRDS(atac_peaks_by_gene, file = file.path(base, dataset, "atac_peaks_by_gene.rds"))

atac_peaks_by_gene <- readRDS(file.path(base, dataset, "atac_peaks_by_gene.rds"))

# compute conservation scores for each gene
gene_results_atac_list <- list()
for (i in seq_along(scent_gene_names)) {gene_results_atac_list[[i]] <- get_gene_phastcons_atac(scent_gene_names[i], atac_peaks_by_gene)}
gene_results_atac <- do.call(rbind, gene_results_atac_list)
saveRDS(gene_results_atac, file = file.path(base, dataset, "gene_phastcons_atac.rds"))
```

```{r}
dataset <- "PBMC" # 10x Genomics

filepath <- file.path(base, dataset, "Pubic-PBMC_allqced_bootpkg_nopc_allCT.FDR0.10.txt")
scent_peaks <- read.table(filepath, header = TRUE, stringsAsFactors = FALSE)
scent_gene_names <- unique(scent_peaks$gene)

## SCENT peaks
scent_peaks_by_gene <- make_scent_peaks_by_gene(scent_peaks)

gene_results_scent_list <- list()
for (i in seq_along(scent_gene_names)) {gene_results_scent_list[[i]] <- get_gene_phastcons_scent(scent_gene_names[i], scent_peaks_by_gene)}
gene_results_scent <- do.call(rbind, gene_results_scent_list)
saveRDS(gene_results_scent, file = file.path(base, dataset, "gene_phastcons_scent.rds"))

## ATAC peaks

# seurat_pbmc <- readRDS(file.path(base, dataset, "seurat_pbmc.rds"))
# atac_peaks <- StringToGRanges(Features(seurat_pbmc[["ATAC"]]), sep = c(":", "-")) # every range has at least one ATAC peak in seurat object
# atac_peaks_by_gene <- make_atac_peaks_by_gene(atac_peaks, scent_gene_names)
# saveRDS(atac_peaks_by_gene, file = file.path(base, dataset, "atac_peaks_by_gene.rds"))

atac_peaks_by_gene <- readRDS(file.path(base, dataset, "atac_peaks_by_gene.rds"))

# compute conservation scores for each gene
gene_results_atac_list <- list()
for (i in seq_along(scent_gene_names)) {gene_results_atac_list[[i]] <- get_gene_phastcons_atac(scent_gene_names[i], atac_peaks_by_gene)}
gene_results_atac <- do.call(rbind, gene_results_atac_list)
saveRDS(gene_results_atac, file = file.path(base, dataset, "gene_phastcons_atac.rds"))
```

```{r}
dataset <- "NeurIPS"
filepath <- file.path(base, dataset, "NeurIPS_allqced_bootpkg_nopc_allCT.FDR0.10.txt")
scent_peaks <- read.table(filepath, header = TRUE, stringsAsFactors = FALSE)
scent_gene_names <- unique(scent_peaks$gene)

## SCENT peaks
scent_peaks_by_gene <- make_scent_peaks_by_gene(scent_peaks)

gene_results_scent_list <- list()
for (i in seq_along(scent_gene_names)) {gene_results_scent_list[[i]] <- get_gene_phastcons_scent(scent_gene_names[i], scent_peaks_by_gene)}
gene_results_scent <- do.call(rbind, gene_results_scent_list)
saveRDS(gene_results_scent, file = file.path(base, dataset, "gene_phastcons_scent.rds"))

## ATAC peaks

seurat_neurips <- readRDS(file.path(base, dataset, "seurat_neurips.rds"))
atac_peaks <- StringToGRanges(Features(seurat_pbmc[["ATAC"]]), sep = c(":", "-")) # every range has at least one ATAC peak in seurat object
atac_peaks_by_gene <- make_atac_peaks_by_gene(atac_peaks, scent_gene_names)
saveRDS(atac_peaks_by_gene, file = file.path(base, dataset, "atac_peaks_by_gene.rds"))

atac_peaks_by_gene <- readRDS(file.path(base, dataset, "atac_peaks_by_gene.rds"))

# compute conservation scores for each gene
gene_results_atac_list <- list()
for (i in seq_along(scent_gene_names)) {gene_results_atac_list[[i]] <- get_gene_phastcons_atac(scent_gene_names[i], atac_peaks_by_gene)}
gene_results_atac <- do.call(rbind, gene_results_atac_list)
saveRDS(gene_results_atac, file = file.path(base, dataset, "gene_phastcons_atac.rds"))
```