# h5ad to Seurat with SeuratDisk
```{r}
library(SeuratDisk)

base <- "/media/chang/HDD-8/joseph/bmi206"

dataset <- "Arthritis"
h5ad_filename <- "GSE243917_ATAC_ALL_CT_processed.h5ad"

h5ad_path <- file.path(base, dataset, h5ad_filename)

options(future.globals.maxSize = 20e+09)

Convert(h5ad_path, dest = "h5seurat", overwrite = FALSE)

h5s <- gsub("\\.h5ad$", ".h5seurat", h5ad_path)
obj <- LoadH5Seurat(h5s, assays = list(RNA = c("counts", "data")))

# saveRDS(obj, file = file.path(base, dataset, 'seurat_arthritis.rds'))
saveRDS(obj, file = file.path(base, dataset, 'seurat_neurips_multiome.rds'))
```

# h5 to Seurat
```{r}
library(Seurat)

h5_path <- file.path(base, dataset, "pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5")
h5_data <- Read10X_h5(h5_path)
metadata <- read.csv(file.path(base, dataset, "pbmc_granulocyte_sorted_10k_per_barcode_metrics.csv"), row.names = 1)

gex_matrix <- h5_data[["Gene Expression"]]
peak_matrix <- h5_data[["Peaks"]]
metadata <- metadata[colnames(gex_matrix), ]

seurat_pbmc <- CreateSeuratObject(counts = gex_matrix, meta.data = metadata)
seurat_pbmc[["ATAC"]] <- CreateAssayObject(counts = peak_matrix)

seurat_pbmc <- subset(seurat_pbmc, subset = nCount_RNA > 1000 & nCount_RNA < 25000 & nCount_ATAC > 1000 & nCount_ATAC < 100000) # QC (supp table 9)

saveRDS(seurat_pbmc, file = file.path(base, dataset, "seurat_pbmc.rds"))
```

# h5ad to Seurat manually (use extract_anndata.ipynb first)
```{r}
library(Seurat)
library(Signac)
library(Matrix)

base <- "/media/chang/HDD-8/joseph/bmi206"
dataset <- "NeurIPS"

expr <- readMM(file.path(base, dataset, "expr.mtx"))
rna_features <- read.delim(file.path(base, dataset, "expr_features.tsv"), sep = "\t", row.names = 1, check.names = FALSE)
rna_cells    <- read.delim(file.path(base, dataset, "expr_cells.tsv"), sep = "\t", row.names = 1, check.names = FALSE)

# expr is cells × genes -> transpose to genes × cells for Seurat
expr <- t(expr)
rownames(expr) <- rownames(rna_features)      # gene IDs
colnames(expr) <- rownames(rna_cells)         # cell barcodes / IDs

seurat_obj <- CreateSeuratObject(counts = expr, meta.data = rna_cells)

atac <- readMM(file.path(base, dataset, "atac.mtx"))
atac_features <- read.delim(file.path(base, dataset, "atac_features.tsv"), sep = "\t", row.names = 1, check.names = FALSE)
atac_cells    <- read.delim(file.path(base, dataset, "atac_cells.tsv"), sep = "\t", row.names = 1, check.names = FALSE)
atac <- t(atac)

rownames(atac) <- rownames(atac_features)      # peak IDs
colnames(atac) <- rownames(atac_cells)         # cell barcodes / IDs

# common_cells <- intersect(colnames(seurat_obj), colnames(atac))
# seurat_obj <- subset(seurat_obj, cells = common_cells)
# atac <- atac[ , common_cells, drop = FALSE]

rownames(atac) <- rownames(atac_features)

chrom_assay <- CreateChromatinAssay(counts = atac,
  sep    = c("-", "-"),  # or c(":", "-")
  genome = 'hg38'
)

seurat_obj[["ATAC"]] <- chrom_assay
saveRDS(seurat_obj, file = file.path(base, dataset, "seurat_neurips.rds"))
```